design/authenticators/authn_sut/authn_sut_solution_design.md

```diff
@@ -0,0 +1,384 @@
+| 12  | Code verifier does not match code challenge | Code verifier does not match code challenge    |
+
+###### Debug Log level
+
+|    | Scenario                                              | Log message                                                                            |
+|--- |-------------------------------------------------------|----------------------------------------------------------------------------------------|
+| 1  | After validating SUT | Single use token for {0-role-name} has been validated successfully                |
+
+##### Audit Log Messages
+
+Same as other authenticators - one message for successful authentication, one for failed authentication.
+
+#### Policy
+
+As with other authenticators, the SUT authenticator will be disabled by default.
+It will be enabled by creating an authenticator policy resource in Conjur and
+adding it to the list of enabled authenticators in the Conjur configuration, for
+example in the `CONJUR_AUTHENTICATORS` environment variable.
+
+The authenticator webservice must be declared in Conjur policy:
+
+```yml
+- !policy
+  id: conjur/authn-sut
+  body:
+  - !webservice
+
+  - !group 
+    id: authenticatable
+    annotations:
+      description: Group with permission to authenticate using this authenticator
+
+  - !permit
+    role: !group authenticatable
+    privilege: [ read, authenticate ]
+    resource: !webservice
+```
+
+and the `authenticatable` group can be used to entitle a user(s) to use the
+authenticator:
+
+```yml
+- !grant
+  role: !group conjur/authn-sut/authenticatable
+  member: !user <username>
+```
+
+#### API
+
+We will need two new API endpoints for the SUT authenticator:
+
+##### `GET /authn-sut/<account>/login`
+
+This endpoint will be used to generate a SUT for a user. The user will provide
+their username and password (or API key) in the `Authentication` header using the
+standard format for HTTP Basic authentication.
+
+Additionally, the client will provide a Code Challenge and Code Challenge Algorithm
+in the headers. For example:
+
+```txt
+Code-Challenge: 3iod...
+Code-Challenge-Algorithm: sha256
+```
+
+For more details on the code challenge, see the [Code Verifier and Code Challenge](#code-verifier-and-code-challenge) section.
+
+The endpoint will return a SUT that can be used to authenticate the user to Conjur.
+
+The format of the URL matches the existing authenticator endpoint for
+authn-ldap.
+
+##### `POST /authn-sut/<account>/authenticate`
+
+This endpoint will be used to authenticate a user to Conjur using a SUT. The
+user will provide the SUT and code verifier in the request body.
+The endpoint will verify that the SUT is valid and has not expired, and that
+the code verifier matches the code challenge associated with the SUT. If this
+is the case, the endpoint will return a short-lived Conjur auth token.
+
+Example request body:
+
+```json
+{
+  "single_use_token": "lkduj...",
+  "code_verifier": "3iod..."
+}
+```
```

Comment:
Mind adding the failure responses as well?

Reply:
Is that necessary? It should be the same as any other authenticator and should not require any custom code.

Reply:
It's really important we capture all the scenarios in these docs. Each authenticator has a specific set of bits that can go wrong that are unique to that authenticator, and it's important we capture that.  

For example, when running in debug mode, I'd expect to see that exchanging an expired token failed because the token was expired. It's helpful to think about that message before we write the implementation to ensure we don't leak security related data in log messages.

Reply:
Only a small portion of authentication failure is handled outside the authenticator.  There's no need to re-document those.  I'm thinking about errors specific to Single-Use tokens. If the above error messages are the final, what do you think about including a debug level and error code for each error? That'll at least give a sense of the expected errors and the communication back to the customer.

Reply:
I'm not seeing where in the code we have custom errors for the other authenticators. Can you please give me some pointers so I have what to compare with?

Reply:
For the record, we discussed this and clarified that the intention is to have more detailed debug logs, not HTTP responses. I'm adding those now.